nvc/extract.go

```diff
@@ -0,0 +1,175 @@
+package nvc
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"os"
+	"strconv"
+)
+
+// TODO: maybe pick a better name for this type
+type compressionType uint32
+
+const (
+	uncompressed compressionType = 0
+	compressed   compressionType = 1
+	encrypted    compressionType = 3
+
+	nvcMagicBytes = "nvc1d\x00\x00\x00"
+)
+
+type nvcEntry struct {
+	Hash        uint64
+	Offset      uint32
+	RawLength   uint32 // Uncompressed length
+	Length      uint32 // Actual length in file
+	Compression compressionType
+}
+
+func Hash2String(hash uint64) string {
+	return fmt.Sprintf("%016x", hash)
+}
+
+func String2Hash(s string) uint64 {
+	hash := fnv.New64a()
+	hash.Write([]byte(s))
+	return hash.Sum64()
+}
+
+func readNVCMagic(r io.Reader) (bool, error) {
+	header := [8]byte{}
+	if _, err := io.ReadFull(r, header[:]); err != nil {
+		return false, err
+	}
+
+	if string(header[:]) != nvcMagicBytes {
+		return false, errors.New(".nvc signature not found")
+	}
+
+	return true, nil
+}
+
+func readNvcEntry(r io.Reader) (nvcEntry, error) {
+	var e nvcEntry
+	if err := binary.Read(r, binary.LittleEndian, &e); err != nil {
+		return e, err
+	}
+	return e, nil
+}
+
+func readCount(r io.Reader) (uint32, error) {
+	var count uint32
+	if err := binary.Read(r, binary.LittleEndian, &count); err != nil {
+		return 0, err
+	}
+	return count, nil
+}
```

Comment:
Is there a reason to have these (and `readNVCMagic`) as separate functions? It might be better to have a single function that takes an `io.Reader` and returns a `([]nvcEntry, error)`. I think that would also reduce code duplication between the two subcommands.

Also, consistency in capitalization in function names (NVC vs Nvc) would be nice (but I really don't care which one we go with as long as we keep it consistent)

Reply:
Since it's a separate package, I'm curious if we could remove the NVC from most of the constants/function names?

I've also gone ahead and combined the readCount and readEntry into one public function, since they probaly are always used together.

Reply:
> Since it's a separate package, I'm curious if we could remove the NVC from most of the constants/function names?

Yeah, good point. `nvc.Entry` does read better than `nvc.nvcEntry`. There's probably something in Effective Go (or something) about that that.

> I've also gone ahead and combined the readCount and readEntry into one public function, since they probaly are always used together.

Right, and at that point, if you need the number of entries.....`len(entries)` :stuck_out_tongue:

Reply:
Yep, found it: https://go.dev/doc/effective_go#package-names

> The importer of a package will use the name to refer to its contents, so exported names in the package can use that fact to avoid repetition...For instance, the buffered reader type in the bufio package is called Reader, not BufReader, because users see it as bufio.Reader, which is a clear, concise name.